src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitUsingTests.cs

```diff
@@ -3084,5 +3084,436 @@ public void Dispose()
+                    static async Task Main()
+                    {
+                        Console.Write("1");
+                        await using (new R())
+                        {
+                            Console.Write("2");
+                        }
+                        Console.Write("3");
+                    }
+                }
+                ref struct R
+                {
+                    public R() => Console.Write("C");
+                    public ValueTask DisposeAsync()
+                    {
+                        Console.Write("D");
+                        return default;
+                    }
+                }
+                """;
+            CreateCompilationWithTasksExtensions(source, parseOptions: TestOptions.Regular12).VerifyDiagnostics(
+                // (8,22): error CS8652: The feature 'ref and unsafe in async and iterator methods' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         await using (new R())
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "new R()").WithArguments("ref and unsafe in async and iterator methods").WithLocation(8, 22));
+
+            var expectedOutput = "1C2D3";
+
+            var comp = CreateCompilationWithTasksExtensions(source, parseOptions: TestOptions.RegularNext, options: TestOptions.ReleaseExe);
+            CompileAndVerify(comp, expectedOutput: expectedOutput).VerifyDiagnostics();
+
+            comp = CreateCompilationWithTasksExtensions(source, options: TestOptions.ReleaseExe);
+            CompileAndVerify(comp, expectedOutput: expectedOutput).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void RefStruct_AwaitUsing_AwaitInside()
+        {
+            var source = """
+                using System.Threading.Tasks;
+                class C
+                {
+                    async Task M()
+                    {
+                        await using (new R())
+                        {
+                            await Task.Yield();
+                        }
+                    }
+                }
+                ref struct R
+                {
+                    public ValueTask DisposeAsync() => default;
+                }
+                """;
+            CreateCompilationWithTasksExtensions(source, parseOptions: TestOptions.Regular12).VerifyDiagnostics(
+                // (6,22): error CS8652: The feature 'ref and unsafe in async and iterator methods' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         await using (new R())
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "new R()").WithArguments("ref and unsafe in async and iterator methods").WithLocation(6, 22));
+
+            var expectedDiagnostics = new[]
+            {
+                // (6,22): error CS4007: Instance of type 'R' cannot be preserved across 'await' or 'yield' boundary.
```

Comment:
>// (6,22): error CS4007: Instance of type 'R' cannot be preserved across 'await' or 'yield' boundary. [](http://example.com/codeflow?start=16&length=101)

It feels like it would be better to report this error across all language versions

Reply:
Right, I will link the issue in a comment. Thanks.